api/src/main/java/ai/djl/training/listener/EarlyStoppingListener.java

```diff
@@ -0,0 +1,175 @@
+package ai.djl.training.listener;
+
+import ai.djl.training.Trainer;
+
+/**
+ * Listener that allows the training to be stopped early if the validation loss is not improving, or if time has expired.
+ * <br/>
+ * Usage:
+ * Add this listener to the training config, and add it as the last one.
+ * <pre>
+ *  new DefaultTrainingConfig(...)
+ *        .addTrainingListeners(new EarlyStoppingListener()
+ *                .setEpochPatience(1)
+ *                .setEarlyStopPctImprovement(1)
+ *                .setMaxMinutes(60)
+ *                .setMinEpochs(1)
+ *        );
+ * </pre>
+ * Then surround the fit with a try catch that catches the {@link EarlyStoppingListener.EarlyStoppedException}.
+ * <br/>
+ * Example:
+ * <pre>
+ * try {
+ *   EasyTrain.fit(trainer, 5, trainDataset, testDataset);
+ * } catch (EarlyStoppingListener.EarlyStoppedException e) {
+ *   // handle early stopping
+ *   log.info("Stopped early at epoch {} because: {}", e.getEpoch(), e.getMessage());
+ * }
+ * </pre>
+ * <br/>
+ * Note: Ensure that Metrics are set on the trainer.
+ */
+public class EarlyStoppingListener implements TrainingListener {
+    private final double objectiveSuccess;
+
+    /**
+     * after minimum # epochs, consider stopping if:
+     */
+    private int minEpochs;
+    /**
+     * too much time elapsed
+     */
+    private int maxMinutes;
```

Comment:
Maybe change this to `maxMillis`. This will let you do the math a bit easier for each epoch. You can still add the helper `setMaxMinutes()` in addition to `setMaxMillis()` in the builder so you can still call it just as easily.

The other benefit for testing is that you can now add a test with it. You can set it to stop after only one ms and then train until timeout.

Reply:
Introduced a test, and made it millis. I also made a convenient method that takes a duration in the builder that converts it for you. I dislike passing big numbers as I usually train for hours.